### PR TITLE
Add runtime/smart contract interactions

### PR DESCRIPTION
**Pull request by justinfrevert**
This PR adds the bulk of the intended changes for this repository - a user-facing example of interactions between the runtime and a smart contract. 

Additions:
- Runtime with a chain extension that exposes a custom extrinsic, and `pallet_balances::transfer` to the smart contract.
- Smart contract with basic storage functions to be called by custom pallet, as well as functions to call chain_extension functions.
- Custom extrinsic that calls pallet_contracts::bare_call with given arguments, to demonstrate calling smart contracts through an extrinsic.

Noting that for the call_smart_contract extrinsic, I've purposefully left out tests and benchmarks due to the amount of additional source needed from pallet_contracts to work with adding the wasm to the mock runtimes - it was a good exercise, but seemed to distract from the focus of this repo.

### CODE REVIEW

File: smart-contracts/example-extension/tests.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+				}
+
+				/// The chain extension is called with the given input.
+				///
+				/// Returns an error code and may fill the `output` buffer with a
+				/// SCALE encoded result. The error code is taken from the
+				/// `ink_env::chain_extension::FromStatusCode` implementation for
+				/// `ContractError`.
+				fn call(&mut self, _input: &[u8], output: &mut Vec<u8>) -> u32 {
+					let ret= $mock_return;
```
Comment (justinfrevert) [inline smart-contracts/example-extension/tests.rs]: This is _really_ static mocking of expected return values. AFAICT, we're not able to do much more than this at the moment.

**Comment by justinfrevert** [comment]
I removed the smart contract tests, in light of our conversations on testing, we're not able to do much more mocking than statically setting the return of runtime functions and it didn't make sense to set an example of doing so. I may play with them a bit on a [separate branch](https://github.com/justinFrevert/Runtime-Contract-Interactions/blob/9504f4fd401527c803e93142431a0c9d0b3e9de6/smart-contracts/example-extension/tests.rs#L51-L81), but not sure it needs to be included here.

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -158,6 +165,83 @@ pub fn native_version() -> NativeVersion {
...
+				env.charge_weight(base_weight.saturating_add(overhead))?;
+
+				let caller = env.ext().caller().clone();
+				let recipient_account = sp_runtime::MultiAddress::Id(recipient);
+
+				pallet_balances::Pallet::<Runtime>::transfer(
+					RawOrigin::Signed(caller).into(),
+					recipient_account,
+					transfer_amount.into(),
+				);
```
Comment (apopiak) [inline runtime/src/lib.rs]: Pretty sure this returns a result.
https://paritytech.github.io/substrate/master/pallet_balances/pallet/struct.Pallet.html#method.transfer

So you probably want to use `?`

Should've also given a warning at compilation, btw.

File: smart-contracts/example-extension/lib.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum ContractError {
+	FailToCallRuntime,
+}
+
+impl From<scale::Error> for ContractError {
+	fn from(_: scale::Error) -> Self {
+		panic!("encountered unexpected invalid SCALE encoding")
```
Comment (apopiak) [inline smart-contracts/example-extension/lib.rs]: This seems very suspect to me. A `From` implementation should not panic

File: smart-contracts/example-extension/lib.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	}
+}
+
+impl ink_env::chain_extension::FromStatusCode for ContractError {
+	fn from_status_code(status_code: u32) -> Result<(), Self> {
+		match status_code {
+			0 => Ok(()),
+			1 => Err(Self::FailToCallRuntime),
+			_ => panic!("encountered unknown status code"),
```
Comment (apopiak) [inline smart-contracts/example-extension/lib.rs]: Similar here, this is runtime code, so it should avoid panicking

File: pallets/template/src/weights.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+/// Weights for pallet_template using the Substrate node and recommended hardware.
+pub struct SubstrateWeight<T>(PhantomData<T>);
+
+impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+
+	
+	
+	// Storage: TemplateModule ContractEntry (r:0 w:1)
+
```
Comment (apopiak) [inline pallets/template/src/weights.rs]: This looks broken formatting-wise

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -158,6 +167,84 @@ pub fn native_version() -> NativeVersion {
...
+		match func_id {
+			// do_store_in_runtime
+			1 => {
+				let mut env = env.buf_in_buf_out();
+				// Capture weight for the main action being performed by the extrinsic
+				let base_weight = RocksDbWeight::get().writes(1);
+				// Add some weight for the contract invocation
+				let overhead = <Runtime as pallet_contracts::Config>::Schedule::get()
+					.host_fn_weights
+					.debug_message;
```
Comment (apopiak) [inline runtime/src/lib.rs]: How is this determined? Was it just chosen randomly to make sure there is some weight?
Why not benchmark and then use the weight of `insert_number`?

**Review by apopiak — COMMENTED** [review]
If this is meant to be a best-practice example I think we should make it as safe as possible and polish e.g. the benchmarking and weight-charging logic.

**Comment by apopiak** [comment]
> Noting that for the call_smart_contract extrinsic, I've purposefully left out tests and benchmarks due to the amount of additional source needed from pallet_contracts to work with adding the wasm to the mock runtimes - it was a good exercise, but seemed to distract from the focus of this repo.

The focus being on interaction of the contract with the underlying runtime?

File: README.md [hunk 7]
```diff
@@ -1,204 +1,137 @@
...
+| :exclamation: This code is not audited or considered ready for production, and should not be used in a production-like environment without any necessary review and changes |
+| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 
-## Getting Started
+## Motivation
 
-Follow the steps below to get started with the Node Template, or get it up and running right from
-your browser in just a few clicks using
-the [Substrate Playground](https://docs.substrate.io/playground/) :hammer_and_wrench:
+Examples of contract-to-runtime interactions are asked about in the context of Substrate support. `Chain extensions` are recommended for this case, and there are [several examples of them available](https://paritytech.github.io/ink-docs/macros-attributes/chain-extension). An example repository demonstrating chain extensions as well as pallet-to-contract calls has not been developed in a recent Substrate version.
```
Comment (apopiak) [inline README.md]: I would focuse the introduction/context on what this repo does.
E.g. some suggested formulation:
```suggestion
This repository provides examples of contract-to-runtime interactions. `ChainExtension`s are recommended for this case, and there are [several examples of them available](https://paritytech.github.io/ink-docs/macros-attributes/chain-extension) from the perspective of usage in `ink!`. This repository aims to detail the Substrate side of `ChainExtension`s to serve as inspiration for future implementations.
```

File: runtime/src/lib.rs (hunk 2 above)
Comment (gautamdhameja) [reply runtime/src/lib.rs]: @apopiak Is there a way to convert `DispatchResultWithPostInfo` to `DispatchResult` or vice versa? I think using the `?` operator here causes the error that says `the trait From<DispatchErrorWithPostInfo<PostDispatchInfo>> is not implemented for `DispatchError`.

Comment (apopiak) [reply runtime/src/lib.rs]:
```suggestion
				pallet_balances::Pallet::<Runtime>::transfer(
					RawOrigin::Signed(caller).into(),
					recipient_account,
					transfer_amount.into(),
				).map(|_| ())?;
```

should work

Comment (apopiak) [reply runtime/src/lib.rs]: ah, got it wrong, it's 
```suggestion
				pallet_balances::Pallet::<Runtime>::transfer(
					RawOrigin::Signed(caller).into(),
					recipient_account,
					transfer_amount.into(),
				).map_err(|e| e.error)?;
```

Comment (justinfrevert) [reply runtime/src/lib.rs]: Yes this works. Thanks

File: runtime/src/lib.rs (hunk 6 above)
Comment (justinfrevert) [reply runtime/src/lib.rs]: We had asked the smart contracts folks about this [here](https://matrix.to/#/!nqwrcufvSwqTNsLMkj:matrix.parity.io/$BJVJ8X61xBl3SrEsPvEoF1WmiDmlEONWpMdlihHcVJ0?via=matrix.parity.io). But I agree that using the weight of `insert_number` ~below~ above would be a more direct way of doing this.

File: README.md (hunk 7 above)
Comment (justinfrevert) [reply README.md]: I can try to update this with this sort of wording, but I think it's important to note that the repository is not meant to only cover contract-to-runtime-interactions, but the inverse of that as well.

Comment (justinfrevert) [reply README.md]: I think if we want to jump into a description straightaway, then perhaps `Motivation` is not the correct title. Still thinking about this. What do you think about introducing the reason _for creating_ this repository?

**Comment by justinfrevert** [comment]
> > Noting that for the call_smart_contract extrinsic, I've purposefully left out tests and benchmarks due to the amount of additional source needed from pallet_contracts to work with adding the wasm to the mock runtimes - it was a good exercise, but seemed to distract from the focus of this repo.
> 
> The focus being on interaction of the contract with the underlying runtime?

I think so. At least that the focus should be on interactions going in either direction. If I'm thinking about going through an example like this as a developer, I would probably end up just copying a lot of code. Something about other developers copying this example largely for one purpose, but getting a veritable library of code for a few tests just _feels_ wrong.

File: README.md (hunk 7 above)
Comment (justinfrevert) [reply README.md]: I updated it to introduce the project summary before this section. I hope that makes sense.

File: smart-contracts/example-extension/lib.rs (hunk 4 above)
Comment (justinfrevert) [reply smart-contracts/example-extension/lib.rs]: I updated this as well, but have one [sanity question](https://matrix.to/#/!tYUCYdSvSYPMjWNDDD:matrix.parity.io/$16388902314572HrVej:matrix.parity.io?via=matrix.parity.io&via=corepaper.org&via=matrix.org) on the subject for the smart contract + ink! folks

Comment (justinfrevert) [reply smart-contracts/example-extension/lib.rs]: Without having a good use for the `1` status code here, I removed it and added a comment explaining why and how a dev would use this in the future.

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -218,7 +219,7 @@ where
 					RawOrigin::Signed(caller).into(),
 					recipient_account,
 					transfer_amount.into(),
-				);
+				).map_err(|dispatch_result| dispatch_result.error)?;
```
Comment (apopiak) [inline runtime/src/lib.rs]: Just FYI: This is a [`DispatchErrorWithPostInfo`](https://github.com/paritytech/substrate/blob/6d09a45caf40a143791deab4f605b95230b787d1/primitives/runtime/src/lib.rs#L506), not a result.

**Review by gautamdhameja — APPROVED** [review]
Looks good to me. Nice progress. Let's merge it after Steve's review and then move it into the dev hub GitHub if possible.

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -158,6 +166,87 @@ pub fn native_version() -> NativeVersion {
...
+				env.write(&result, false, None)
+					.map_err(|_| "Encountered an error when querying balance.")?;
+			},
+			// do_get_from_runtime
+			4 => {
+				let mut env = env.buf_in_buf_out();
+				let result = TemplateModule::get_value().encode();
+				env.write(&result, false, None)
+					.map_err(|_| "Encountered an error when retrieving runtime storage value.")?;
+			},
```
Comment (apopiak) [inline runtime/src/lib.rs]: Note that these don't charge any weight. I'm unsure whether that's correct.

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -158,6 +166,87 @@ pub fn native_version() -> NativeVersion {
 
 const NORMAL_DISPATCH_RATIO: Perbill = Perbill::from_percent(75);
 
+impl ChainExtension<Runtime> for ExampleExtension
```
Comment (stiiifff) [inline runtime/src/lib.rs]: I would move the extension to a separate file.

File: runtime/src/lib.rs [hunk 11]
```diff
@@ -158,6 +166,87 @@ pub fn native_version() -> NativeVersion {
...
+				let base_weight: Weight =
+					<Runtime as pallet_template::Config>::WeightInfo::insert_number(value);
+				// Add some weight for the contract invocation
+				let overhead = <Runtime as pallet_contracts::Config>::Schedule::get()
+					.host_fn_weights
+					.debug_message;
+				env.charge_weight(base_weight.saturating_add(overhead))?;
+				let caller = env.ext().caller().clone();
+
+				crate::pallet_template::Pallet::<Runtime>::insert_number(
```
Comment (stiiifff) [inline runtime/src/lib.rs]: All calls to the runtime are strongly-coupled with the target pallet(s). I think it's fine for a first iteration, but could be interesting to make those calls loosely-coupled (e.g. via traits) at a later point.

**Review by stiiifff — COMMENTED** [review]
Moreover, you could run cargo fmt & clippy:
- `cargo +nightly fmt --all`
- `cargo clippy -- -D warnings -A deprecated`

File: runtime/src/lib.rs [hunk 12]
```diff
@@ -176,22 +176,22 @@ where
 		E: Ext<T = Runtime>,
 		<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
 	{
+		let mut env = env.buf_in_buf_out();
+		let contracts_overhead = <Runtime as pallet_contracts::Config>::Schedule::get()
+			.host_fn_weights
+			.debug_message;
```
Comment (apopiak) [inline runtime/src/lib.rs]: :+1:  I would also add a comment with some explanation

File: runtime/src/lib.rs (hunk 9 above)
Comment (justinfrevert) [reply runtime/src/lib.rs]: Discussed in DM and committed changes with these weights

File: runtime/src/lib.rs (hunk 11 above)
Comment (justinfrevert) [reply runtime/src/lib.rs]: Yes, it seems like a good task to follow-on

File: runtime/src/lib.rs (hunk 12 above)
Comment (justinfrevert) [reply runtime/src/lib.rs]: I took a swing at it

**Review by stiiifff — APPROVED** [review]

**Pull request merged**